inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/fully_connected/fully_connected_kernel_bf_tiled.cpp

```diff
@@ -0,0 +1,337 @@
+    if (!Parent::Validate(params, options))
+        return false;
+
+    auto& fc_params = static_cast<const fully_connected_params&>(params);
+    auto& input = fc_params.inputs[0];
+
+    // Block reads must be aligned to 4 bytes, for fp16 we can correct for offset misalignment,
+    // but we need to ensure that batch pitch preserves alignment.
+    if (input.GetDType() == Datatype::F16 && input.Batch().pitch % 2 != 0 && input.Batch().v > 1)
+        return false;
+
+    if (input.GetLayout() == DataLayout::bfyx) {
+        // Padding on input is not supported.
+        // TODO: Enable by mirroring the padding in weights.
+        if (input.X().pad.Total() != 0)
+            return false;
+        if (input.Y().pad.Total() != 0)
+            return false;
+    }
+
+    return true;
+}
+
+namespace {
+
+struct TuneParamsSelector {
+    using tune_params = FullyConnected_bf_tiled::tune_params;
+    using functional_case = std::function<tune_params(const fully_connected_params&)>;
+
+    TuneParamsSelector(const fully_connected_params& params) : params(params), selected(false) {}
+
+    TuneParamsSelector& Case(const tune_params& tparams) {
+        if (!selected && VerifyTuneParams(params, tparams)) {
+            result = tparams;
+            selected = true;
+        }
+        return *this;
+    }
+
+    TuneParamsSelector& Case(functional_case fun) {
+        return Case(fun(params));
+    }
+
+    tune_params Default(const tune_params& tparams) {
+        if (!selected) {
+            selected = true;
+            result = tparams;
+        }
+        return result;
+    }
+
+    tune_params Default(functional_case fun) {
+        return Default(fun(params));
+    }
+
+    static bool VerifyTuneParams(const fully_connected_params& params, const tune_params& tparams);
+
+    const fully_connected_params& params;
+    bool selected;
+    tune_params result;
+};
+
+bool TuneParamsSelector::VerifyTuneParams(const fully_connected_params& params, const tune_params& tparams) {
+    // Check divisibility by dispatch tile sizes.
+    if (params.output.Batch().v % (tparams.tile_b * tparams.dispatch_bsv) != 0)
+        return false;
+    if (CeilDiv(params.output.Feature().v, tparams.tile_ofm * simd) % tparams.dispatch_fsv != 0)
+        return false;
+
+    // Same result can be achieved with smaller tile_ofm.
+    if (params.output.Feature().v <= (tparams.tile_ofm / 2) * simd)
+        return false;
+    // No weights layout for such huge tile ofm.
+    if (tparams.tile_ofm * simd > 64)
+        return false;
+
+    // Reject tile sizes that are guaranteed to spill out of registers.
+    unsigned acc_register_bytes = tparams.tile_b * tparams.tile_ofm * simd * BytesPerElement(params.inputs[0].GetDType());
+    unsigned in_register_bytes = tparams.tile_b * tparams.tile_ifm * simd * BytesPerElement(params.inputs[0].GetDType());
+    unsigned wei_register_bytes = tparams.tile_ofm * tparams.tile_k * simd * BytesPerElement(params.weights.GetDType());
```

Comment:
What about the calculation of register bytes for additional variables, indices, etc.? Is this size very small?

Reply:
I haven't had time to estimate that properly. 
Also default selection keeps the registers count in range where they are guaranteed not to spill, so this can only come out during auto-tuning.